pkg/client/client.go

```diff
@@ -37,66 +36,139 @@ import (
-		secretInformer:    secretInformer,
-		configMapInformer: cmInformer,
-	}, nil
+		secretInformer:                    secretInformer,
+		configMapInformer:                 cmInformer,
+		managerNutanixPrismEndpointReader: readManagerNutanixPrismEndpointFromFile,
+	}
+}
+
+func (n *NutanixClientHelper) withCustomNutanixPrismEndpointReader(getter func() (*credentials.NutanixPrismEndpoint, error)) *NutanixClientHelper {
+	n.managerNutanixPrismEndpointReader = getter
+	return n
 }
 
-func (n *NutanixClientHelper) GetClientFromEnvironment(ctx context.Context, nutanixCluster *infrav1.NutanixCluster) (*nutanixClientV3.Client, error) {
+// BuildClientForNutanixClusterWithFallback builds a Nutanix Client from the information provided in nutanixCluster.
+func (n *NutanixClientHelper) BuildClientForNutanixClusterWithFallback(ctx context.Context, nutanixCluster *infrav1.NutanixCluster) (*nutanixClientV3.Client, error) {
+	me, err := n.buildManagementEndpoint(ctx, nutanixCluster)
+	if err != nil {
+		return nil, err
+	}
+	creds := prismgoclient.Credentials{
+		URL:      me.Address.Host,
+		Endpoint: me.Address.Host,
+		Insecure: me.Insecure,
+		Username: me.ApiCredentials.Username,
+		Password: me.ApiCredentials.Password,
+	}
+	return Build(creds, me.AdditionalTrustBundle)
+}
+
+// buildManagementEndpoint takes in a NutanixCluster and constructs a ManagementEndpoint with all the information provided.
+// If required information is not set, it will fallback to using information from /etc/nutanix/config/prismCentral,
+// which is expected to be mounted in the Pod.
+func (n *NutanixClientHelper) buildManagementEndpoint(ctx context.Context, nutanixCluster *infrav1.NutanixCluster) (*envTypes.ManagementEndpoint, error) {
 	log := ctrl.LoggerFrom(ctx)
-	// Create a list of env providers
+
+	// Create an empty list of env providers
 	providers := make([]envTypes.Provider, 0)
 
-	// If PrismCentral is set, add the required env provider
-	prismCentralInfo := nutanixCluster.Spec.PrismCentral
-	if prismCentralInfo != nil {
-		if prismCentralInfo.Address == "" {
-			return nil, ErrPrismAddressNotSet
-		}
-		if prismCentralInfo.Port == 0 {
-			return nil, ErrPrismPortNotSet
-		}
-		credentialRef, err := GetCredentialRefForCluster(nutanixCluster)
-		if err != nil {
-			//nolint:wrapcheck // error is alredy wrapped
-			return nil, err
-		}
-		// If namespace is empty, use the cluster namespace
-		if credentialRef.Namespace == "" {
-			credentialRef.Namespace = nutanixCluster.Namespace
-		}
-		additionalTrustBundleRef := prismCentralInfo.AdditionalTrustBundle
-		if additionalTrustBundleRef != nil &&
-			additionalTrustBundleRef.Kind == credentialTypes.NutanixTrustBundleKindConfigMap &&
-			additionalTrustBundleRef.Namespace == "" {
-			additionalTrustBundleRef.Namespace = nutanixCluster.Namespace
-		}
-		providers = append(providers, kubernetesEnv.NewProvider(
-			*nutanixCluster.Spec.PrismCentral,
-			n.secretInformer,
-			n.configMapInformer))
```

Comment:
I pulled this out into its own function to make the codepaths more testable, the logic did not change.